### PR TITLE
[Core] Disable conda activate for custom docker image as runtime environment

### LINKED ISSUE #3741: [Core] SkyPilot conda shadows pre-configured environment in docker containers
When using docker containers as the runtime environment (`image_id: docker:...`), SkyPilot's installed conda gets activated by default during `run`, `setup` and when you ssh. This trips up users, who may have configured their envs without conda:

Examples include nemo and vllm images:
```
sky launch --image-id docker:vllm/vllm-openai:latest

# python -m vllm.entrypoints.openai.api_server does not work without conda deactivate
```

Our vLLM yaml also does the same: https://github.com/skypilot-org/skypilot/blob/38b101d8949a668ff13284313e3d92a79287e3a6/llm/vllm/serve-openai-api-docker.yaml#L12

More quotes from users:
> vllm image doesn’t work super well as a base image for skypilot.  Doesn’t seem to have anything useful installed
> ...
> explains why my past attempts to use this approach all failed

Another user runs on k8s with custom image. He is forced to runs conda deactivate in the YAMLs he shared with us.

We should probably not activate conda by default when docker image is specified.

### PR DESCRIPTION
**Pull request by cblmemo**
Blocked by #3867 and fixes #3741.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
     - [x] vLLM docker image works:
```yaml
resources:
  cloud: aws
  image_id: docker:vllm/vllm-openai:latest
  accelerators: T4
run: |
  # conda deactivate # <- dont need to deavtivate
  python3 -c "import huggingface_hub; huggingface_hub.login('hf_xxx')"
  python3 -m vllm.entrypoints.openai.api_server --model meta-llama/Llama-2-7b-chat-hf --tokenizer hf-internal-testing/llama-tokenizer --max-model-len 64
```
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests:
     - [x] `pytest tests/test_smoke.py::test_docker_storage_mounts`
     - [x] `pytest tests/test_smoke.py::test_job_queue_with_docker`
     - [x] `pytest tests/test_smoke.py::test_docker_preinstalled_package` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/skylet/constants.py [hunk 1]
```diff
@@ -126,7 +126,7 @@
 # https://github.com/ray-project/ray/issues/31606
 # We use python 3.10 to be consistent with the python version of the
 # AWS's Deep Learning AMI's default conda environment.
-CONDA_INSTALLATION_COMMANDS = (
+CONDA_INSTALLATION_COMMANDS = lambda conda_auto_activate: (
```
Comment (Michaelvll) [inline sky/skylet/constants.py:129]: nit: Can we just use `.format` for this purpose? Having a lambda function in the `constants.py` is a bit weird.

Comment (cblmemo) [reply sky/skylet/constants.py:129]: I actually tried, but `.format` does not work for those `{}` in the installation commands..

**Comment by cblmemo** [comment]
Smoke test for this also passed!

File: sky/skylet/constants.py (hunk 1 above)
Comment (Michaelvll) [reply sky/skylet/constants.py:129]: I see, in that case, let's have a comment here for why we use `lambda` function here. : )

**Review by Michaelvll — APPROVED** [review]

**Comment by scm-aiml** [comment]
If I have a custom built docker container that is store in gcp and that is where I pull from. How do I enable this feature to prevent conda from activating?

**Comment by cg505** [comment]
@scm-aiml You can specify the docker image in `resources.image_id` in your SkyPilot YAML: https://docs.skypilot.co/en/latest/reference/yaml-spec.html#yaml-spec-resources-image-id

**Comment by scm-aiml** [comment]
@cg505 think there might be some confusion. I am specifying my image in that resource location and my image is being pulled, which is hosted on google cloud container registry. I am still seeing it enable base conda on the run section, and I'm having to do 'conda deactivate'. 

I'm not sure if it's because it isn't pulling directly from docker that it's not trigger the flag, or if I'm missing something. 

We're on v1.0.0.dev20250405

**Comment by cg505** [comment]
This sounds like a bug then. I created an issue.

**Pull request merged**